NumLib/Function/Interpolation.h

```diff
@@ -16,38 +16,69 @@
 namespace NumLib
 {
 
-/**
+namespace detail
+{
+
+template<unsigned DOFOffset, typename NodalValues, typename ShapeMatrix>
+void shapeFunctionInterpolate(
+        const NodalValues &/*nodal_values*/,
+        const ShapeMatrix &/*shape_matrix_N*/)
+{}
+
+template<unsigned DOFOffset, typename NodalValues, typename ShapeMatrix, typename... Numbers>
+void shapeFunctionInterpolate(
+        const NodalValues &nodal_values,
+        const ShapeMatrix &shape_matrix_N,
+        double& interpolated_value,
+        Numbers&... interpolated_values)
+{
+    auto const num_nodes = shape_matrix_N.size();
+    double iv = 0.0;
+
+    for (auto n=decltype(num_nodes){0}; n<num_nodes; ++n) {
+        iv += nodal_values[DOFOffset*num_nodes+n] * shape_matrix_N[n];
+    }
+
+    interpolated_value = iv;
+
+    shapeFunctionInterpolate<DOFOffset+1>(nodal_values, shape_matrix_N, interpolated_values...);
+}
+
+} // namespace detail
+
+/*!
  * Interpolates variables given at element nodes according to the given shape matrix.
  *
  * This function simply does the usual finite-element interpolation, i.e. multiplication
  * of nodal values with the shape function.
  *
- * @param nodal_values   vector of nodal values, ordered by component
- * @param shape_matrix_N shape matrix of the point to which will be interpolated
- * @param interpolated_values array of addresses to which the interpolated values will be written
+ * \param nodal_values vector of nodal values, ordered by component
+ * \param shape_matrix_N shape matrix of the point to which will be interpolated
+ * \param interpolated_value interpolated value of the first d.o.f. (output parameter)
+ * \param interpolated_values interpolated value of further d.o.f. (output parameter)
+ *
+ * \note
+ * \c nodal_values have to be ordered by component and it is assumed that all passed d.o.f. are
+ * single-component and are interpolated using the same shape function.
  */
-template<typename NodalValues, typename ShapeMatrix, std::size_t NodalDOF>
+template<typename NodalValues, typename ShapeMatrix, typename... Numbers>
 void shapeFunctionInterpolate(
         const NodalValues& nodal_values,
         const ShapeMatrix& shape_matrix_N,
-        std::array<double*, NodalDOF> interpolated_values
+        double& interpolated_value,
+        Numbers&... interpolated_values
         )
 {
+    auto const num_nodal_dof = sizeof...(interpolated_values) + 1;
```

Comment:
Variadic arguments without variadic types would be the best solution. Let's ask Herb Sutter, if he can do anything about that. :smile: